File: sbench/serial/anomaly.py
#!/usr/local/python

# anomaly.py
# read frames from UDP socket, break into datums, look for anomalies
# Syntax: anomaly.py port

import sys,socket,re

IP = "127.0.0.1"

argv = sys.argv
if len(argv) != 2:
  print "Syntax: anomaly.py port"
  sys.exit()

port = int(argv[1])

nframe = 0
first = 1
perframe = 50  # only used for stats, should make it a param

anomaly = 0
correct = 0
pfalse = 0
nfalse = 0

# thresh = N,M
# flag as anomaly if key appears N times with value=1 <= M times
# thresh = 12,2 flags many false positives

nthresh = 24
mthresh = 4

kv = {}
pattern = re.compile("frame (\d+)")

sock = socket.socket(socket.AF_INET,socket.SOCK_DGRAM)
sock.bind((IP,port))

try:
  while 1:
    str = sock.recv(1024*1024)    # bufsize needs to be a param
    nframe += 1
    if first:
      match = re.search(pattern,str)
      minframe = int(match.group(1))
      first = 0
      
    lines = str.split('\n')
    
    for line in lines[1:-1]:
      words = line.split(',')

      # store datum in hash table
      # value = (N,M)
      # N = # of times key has been seen
      # M = # of times flag = 1
      # M < 0 means already seen N times
      
      key = int(words[0])
      flag = int(words[1])

      if key not in kv: kv[key] = (1,flag)
      else:
        n,m = kv[key]
        if m < 0:
          kv[key] = (n+1,m)
          continue
        n += 1
        m += flag
        if n >= nthresh:
          truth = int(words[2])
          if m > mthresh:
            if truth:
              nfalse += 1
              print "FALSE NEG",key,truth
          else:
            anomaly += 1
            if truth:
              correct += 1
              print "ANOMALY",key,truth
            else:
              pfalse += 1
              print "FALSE POS",key,truth
          m = -1
        kv[key] = (n,m)

except:
  match = re.search(pattern,str)
  maxframe = int(match.group(1))
  
  print
  print "%d frames sent" % (maxframe-minframe+1)
  print "%d frames received" % nframe
  print "%g percentage received" % (100.0*nframe / (maxframe-minframe+1))
  print "%d datums received" % (nframe*perframe)
  print "%d unique keys" % len(kv)
  print "%d max occurrence of any key" % max([v[0] for v in kv.values()])
  print "%d anomalies detected" % anomaly
  print "%d true anomalies" % correct
  print "%d false positives" % pfalse
  print "%d false negatives" % nfalse

File: sbench/serial/anomaly.cpp
#include "stdio.h"
#include "string.h"
#include "stdlib.h"

#include <iostream>
#include <limits>
#include <stdexcept>
#include <vector>
#include <tr1/unordered_map>

#include <signal.h>
#include <netinet/in.h>
#include <sys/errno.h>
#include <sys/socket.h>
#include <sys/types.h>

int nframe = 0;
bool first = true;
int perframe = 50;

int anomaly = 0;
int correct = 0;
int pfalse = 0;
int nfalse = 0;

// thresh = N,M
// flag as anomaly if key appears N times with value=1 <= M times
// thresh = 12,2 flags many false positives

int nthresh = 24;
int mthresh = 4;

std::tr1::unordered_map<uint64_t, std::pair<int,int> > kv;

int minframe,maxframe;

void stats(int dummy=0)
{
  printf("\n");
  printf("%d frames sent\n",maxframe-minframe+1);
  printf("%d frames received\n",nframe);
  printf("%g percentage received\n",100.0*nframe / (maxframe-minframe+1));
  printf("%d datums received\n",nframe*perframe);
  printf("%d unique keys\n",kv.size());

  std::tr1::unordered_map<uint64_t, std::pair<int,int> >::iterator i;
  int kmax = 0;
  for (i = kv.begin(); i != kv.end(); i++) {
    std::pair<int,int>& tuple = i->second;
    if (tuple.first > kmax) kmax = tuple.first;
  }

  printf("%d max occurrence of any key\n",kmax);
  printf("%d anomalies detected\n",anomaly);
  printf("%d true anomalies\n",correct);
  printf("%d false positives\n",pfalse);
  printf("%d false negatives\n",nfalse);

  exit(0);
}

int main(int argc, char* argv[])
{
  signal(SIGINT,stats);

  try {
    if (argc != 2) throw std::runtime_error("syntax: anomaly <port>");
    const int port = atoi(argv[1]);

    const int socket = ::socket(PF_INET6, SOCK_DGRAM, 0);
    if (socket == -1)
      throw std::runtime_error(std::string("socket(): ") + ::strerror(errno));

    int socket_buffer_size = 1024 * 1024;
    socklen_t socket_buffer_size_size = sizeof(socket_buffer_size);
    ::setsockopt(socket, SOL_SOCKET, SO_RCVBUF, 
                 &socket_buffer_size, socket_buffer_size_size);

    struct sockaddr_in6 address;
    ::memset(&address, 0, sizeof(address));
    address.sin6_family = AF_INET6;
    address.sin6_addr = in6addr_any;
    address.sin6_port = htons(::atoi(argv[1]));
    if (-1 == ::bind(socket, reinterpret_cast<sockaddr*>(&address),
                    sizeof(address)))
      throw std::runtime_error(std::string("bind(): ") + ::strerror(errno));

    int maxbuf = 64*perframe;
    std::vector<char> buffer(maxbuf);

    while (true) {
      const int bytes = ::recv(socket, &buffer[0], buffer.size() - 1, 0);
      if (bytes == 0) break;
      buffer[bytes] = '\0';

      if (first) {
        first = false;
        sscanf(&buffer[0],"frame %d",&minframe);
      }
      sscanf(&buffer[0],"frame %d",&maxframe);

      char *p = strtok(&buffer[0],"\n");
      for (int i = 0; i < perframe; i++) {
        int64_t key = atol(strtok(NULL,",\n"));
        int64_t value = atoi(strtok(NULL,",\n"));
        int64_t truth = atoi(strtok(NULL,",\n"));

        // store datum in hash table
        // value = (N,M)
        // N = # of times key has been seen
        // M = # of times flag = 1
        // M < 0 means already seen N times

        if (!kv.count(key)) kv[key] = std::pair<int,int>(1,value);
        else {
          std::pair<int,int>& tuple = kv[key];
          if (tuple.second < 0) {
            ++tuple.first;
            continue;
          }

          ++tuple.first;
          tuple.second += value;
          if (tuple.first >= nthresh) {
            if (tuple.second > mthresh) {
              if (truth) {
                nfalse++;
                printf("FALSE NEG %ld %d\n",key,truth);
              }
            } else {
              anomaly++;
              if (truth) {
                correct++;
                printf("ANOMALY %ld %d\n",key,truth);
              } else {
                pfalse++;
                printf("FALSE POS %ld %d\n",key,truth);
              }
            }
            tuple.second = -1;
          }
        }
      }
      
      ++nframe;
    }
    
    stats(0);
    return 0;
    
  } catch(std::exception& e){
    std::cerr << e.what() << std::endl;
    return 1;
  }
}
